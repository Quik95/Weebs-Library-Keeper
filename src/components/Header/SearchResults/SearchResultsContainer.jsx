import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useManualQuery } from "graphql-hooks";
import { get } from "lodash";
import clsx from "clsx";

// MUI
import List from "@material-ui/core/List";
import CircularProgress from "@material-ui/core/CircularProgress";

// styles
import useStyles from "./SearchResults.styles";

// custom components
import ListItem from "./ListItem";

// GraphQL queries
import QUERY from "../../../queries/searchSeries";

export default function SearchResultsContainer({ searchTerm }) {
  const [shouldBeOpen, setShouldBeOpen] = useState(false);
  const [fetchResults, { loading, error, data }] = useManualQuery(QUERY, {
    variables: { query: searchTerm }
  });
  const classes = useStyles();

  const closeResults = () => setShouldBeOpen(false);

  const hideOnOutOfBoundsClick = e => {
    const path = e.path;
    const wasSearchResultsContainerInPath = path.find(entry => {
      return get(entry, "dataset.testid") === "SearchContainer";
    });
    if (wasSearchResultsContainerInPath) return;
    setShouldBeOpen(false);
  };

  useEffect(() => {
    const root = document.querySelector("html");
    root.addEventListener("click", hideOnOutOfBoundsClick);

    if (!searchTerm) setShouldBeOpen(false);
    else {
      setShouldBeOpen(true);
      fetchResults();
    }
    return () => {
      root.removeEventListener("click", hideOnOutOfBoundsClick);
    };
  }, [fetchResults, searchTerm]);

  if (!shouldBeOpen) return null;
  if (error) console.error(error.messege);
  if (loading)
    return (
      <List
        className={clsx(classes.listContainer, classes.spinnerFix)}
        data-testid="SearchContainer">
        <CircularProgress size={40} variant="indeterminate" color="primary" />
      </List>
    );
  if (data) {
    const resList = get(data, "Page.media", []);
    return (
      <List className={classes.listContainer} data-testid="SearchContainer">
        {resList.map(anime => (
          <ListItem key={anime.id} data={anime} handleClose={closeResults} />
        ))}
      </List>
    );
  }
}

SearchResultsContainer.propTypes = {
  searchTerm: PropTypes.string
};
