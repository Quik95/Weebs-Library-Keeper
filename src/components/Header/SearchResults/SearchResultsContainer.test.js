import SearchResultsContainer from './SearchResultsContainer';

describe('<SearchResultsContainer/>', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<SearchResultsContainer />);
    expect(wrapper).toMatchSnapshot();
  });
});
