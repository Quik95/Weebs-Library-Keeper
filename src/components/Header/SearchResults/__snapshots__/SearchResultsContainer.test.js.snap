// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultsContainer/> Renders correctly 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-listContainer-1"
  data-testid="SearchContainer"
>
  <Memo(ListItem)
    key="0"
  />
  <Memo(ListItem)
    key="1"
  />
  <Memo(ListItem)
    key="2"
  />
  <Memo(ListItem)
    key="3"
  />
  <Memo(ListItem)
    key="4"
  />
  <Memo(ListItem)
    key="5"
  />
  <Memo(ListItem)
    key="6"
  />
  <Memo(ListItem)
    key="7"
  />
  <Memo(ListItem)
    key="8"
  />
  <Memo(ListItem)
    key="9"
  />
  <Memo(ListItem)
    key="10"
  />
</WithStyles(ForwardRef(List))>
`;
